### DISCUSSION TITLE
Listening for hash changes in custom hook

### DISCUSSION THREAD

**Discussion by dominik-ilja** [vercel/next.js#64463] [Help]
### Summary

I'm trying to implement a simple custom hook to listen for URL hash changes in v14. The `"hashchange"` event doesn't trigger at all when using Next.

### Additional information

```js
"use client";

import { useState, useEffect } from "react";

export const useHash = () => {
  const [hash, setHash] = useState("");

  useEffect(() => {
    function onHashChange() {
      setHash(window.location.hash);
    }

    const event = "hashchange";
    window.addEventListener(event, onHashChange);
    return window.removeEventListener(event, onHashChange);
  }, []);

  return hash;
};
```
```


### Example

_No response_

**Comment by Binney** [comment]
Hi @dominik-ilja - found your thread while trying to do the same thing. Your issue is you're adding the event listener and then immediately removing it again. The return value of `useEffect` should be a function, rather than calling the function.

If you change it to

```
"use client";

import { useState, useEffect } from "react";

export const useHash = () => {
  const [hash, setHash] = useState("");

  useEffect(() => {
    function onHashChange() {
      setHash(window.location.hash);
    }

    const event = "hashchange";
    window.addEventListener(event, onHashChange);
    return () => window.removeEventListener(event, onHashChange);
  }, []);

  return hash;
};
```

this works as expected.

A few other sources have said this too but this isn't really a NextJS thing, just a vanilla client-side React + Javascript thing. AFAICT Next doesn't provide much extra for URL hashes, since they only work on the client and you can use standard React for them.
